### PR TITLE
Evaluator skeleton

### PR DESCRIPTION
**Pull request by thomastanck**

### CODE REVIEW

**Comment by chrisjwelly** [comment]
@thomastanck There is a conflicting file in this PR since #4 was merged

File: src/evaluator/__tests__/evaluator.ts [hunk 1]
```diff
@@ -1,9 +1,138 @@
...
+
+  test('invalid', () => {
+    // too few args
+    expectJsonReadEvalError(['quote'], undefined).toMatchInlineSnapshot(`undefined`);
+    // too many args
+    expectJsonReadEvalError(['quote', 'hey', 'there'], undefined).toMatchInlineSnapshot(
+      `undefined`
+    );
+    // improper list
+    expectJsonReadEvalError(['quote', '.', 'a'], undefined).toMatchInlineSnapshot(`undefined`);
```
Comment (kwannoel) [inline src/evaluator/__tests__/evaluator.ts:80]: Doesn't this evaluate to a pair?

Comment (kwannoel) [reply src/evaluator/__tests__/evaluator.ts:80]: Rather than `undefined`

File: src/reader/reader.ts [hunk 2]
```diff
@@ -41,7 +41,7 @@ function readList(firsttok: Tok.LPar, tokens: Token[], i: number): PartialReadRe
   const data3: SExpr[] = [];
 
   // Stage 1, no . seen yet
-  while (true) {
+  for (;;) {
```
Comment (kwannoel) [inline src/reader/reader.ts:44]: didn't know you could do this! Is this a code style change or is there a semantic difference with `while (true)`?

File: src/sexpr/sexpr.ts [hunk 3]
```diff
@@ -8,6 +8,7 @@ export enum STypes {
   Boolean,
   Nil,
   List,
+  Boxed,
```
Comment (kwannoel) [inline src/sexpr/sexpr.ts:11]: This is referencing: https://docs.racket-lang.org/reference/boxes.html right? Why do we include this?

File: src/evaluator/__tests__/evaluator.ts (hunk 1 above)
Comment (thomastanck) [reply src/evaluator/__tests__/evaluator.ts:80]: yes, which is why it's an error

File: src/reader/reader.ts (hunk 2 above)
Comment (thomastanck) [reply src/reader/reader.ts:44]: eslint doesn't like constant conditions, so this is just another way of writing an infinite loop that doesn't have constant conditions

File: src/sexpr/sexpr.ts (hunk 3 above)
Comment (thomastanck) [reply src/sexpr/sexpr.ts:11]: no actually it wasn't, i guess i should rename it :thinking: 

it's supposed to be a leaf node that can extend the s expression type easily to support other kinds of leaves. by default, s expressions only support 3 leaf nodes, atom, boolean, and number.

Comment (kwannoel) [reply src/sexpr/sexpr.ts:11]: What other types of leaf nodes would we want to support with this?

**Review by kwannoel — APPROVED** [review]

**Pull request merged**